### PR TITLE
#174797522 Profile Page Settings

### PR DESCRIPTION
**Pull request by ericus123**
**What does this PR do?**
 This PR complete , display and update user profile details.
**Description of Task to be completed?**

  - Complete user profile by providing additional details after successful registration 
  - Validate user profile Completion.
  - Display user profile
  - Update user profile 
  - Validate user profile update
  - Document profile page settings feature using swagger
    
**How should this be manually tested?**

  - Clone this repo `git clone https://github.com/atlp-rwanda/script-heroes-bn-backend.git`
  - Switch your working directory by running` cd script-heroes-bn-backend`
  - Checkout the branch `git checkout ft-profile-page-settings-#174797522`
  - Open the folder with you code editor
  - Start the server
  - Open postman and test using the models provided below.

`To get access you need to set request headers {'x-auth-token':'token of logged in user'}`

1. Completing user profile` http://localhost:3000/api/profile/complete`
 ```
  {
       "gender":"Male",
      "birthdate":"1978-05-04",
      "language":"Kinyarwanda",
      "currency":"US$",
      "country":"Rwanda",
      "department":"IT",
      "linemanager":12
  }
```

2. Displaying user profile  `http://localhost:3000/api/profile`
3. Updating user profile` http://localhost:3000/api/profile/update`
```
    {
        "firstName": "name",
        "lastName": "lastName",
        "phoneNumber": "12345678",
        "gender": "Male",
        "birthdate": "1978-05-04T00:00:00.000Z",
        "language": "Kinyarwanda",
        "currency": "US$",
        "country": "Rwanda",
        "department": "IT",
        "linemanager": 12
    }
```

**What are the relevant pivotal tracker stories?**
[   #174797522](#174797522)
**Screenshots (if appropriate)**

Screenshot from 2020-10-08 11-49-41
Screenshot from 2020-10-08 11-51-00
Screenshot from 2020-10-08 11-51-33

### CODE REVIEW

File: src/database/migrations/20200928102122-create-user.js [hunk 1]
```diff
@@ -22,6 +22,30 @@ module.exports = {
...
+      },
+      country: {
+        type: Sequelize.STRING
+      },
+      role: {
+        type: Sequelize.STRING
+      },
+      department: {
+        type: Sequelize.STRING
+      },
```
Comment (kipropbrian) [inline src/database/migrations/20200928102122-create-user.js]: Hey, remember to not touch old migrations or rename them. You have to create new migrations for each feature.

File: src/utils/internationalization/locales/en.json [hunk 2]
```diff
@@ -1,36 +1,64 @@
...
+  "Role must be atleast {#limit} characters long": "Role must be atleast {#limit} characters long",
+  "Role is required": "Role is required",
+  "Department must be a string": "Department must contain letters only",
+  "Please fill in your department": "Please fill in your department",
+  "Department must be atleast {#limit} characters long": "Department must be atleast {#limit} characters long",
+  "Department is required": "Department is required",
+  "Line manager must be a string": "Line manager must contain letters only",
+  "Please fill in the line manager": "Please fill in the line manager",
+  "Line manager must be atleast {#limit} characters long": "Line manager must be atleast {#limit} characters long",
+  "Line manager is required": "Line manager is required"
```
Comment (elvisiraguha) [inline src/utils/internationalization/locales/en.json]: Could these be translated too?

**Review by elvisiraguha — COMMENTED** [review]
Kindly rebase

File: src/utils/internationalization/locales/en.json (hunk 2 above)
Comment (ericus123) [reply src/utils/internationalization/locales/en.json]: They are tlanslated now

File: src/middlewares/verifyToken.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+export default (req, res, next) => {
+  const token = req.header('auth-token');
+  if (!token) return res.status(401).send({ error: 'Acces Denied' });
+  try {
+    const verified = jwt.verify(token, process.env.JWT_SECRET);
+    req.user = verified;
+    next();
+  } catch (err) {
+    res.status(400).send({ error: 'Invalid Token' });
+  }
```
Comment (elvisiraguha) [inline src/middlewares/verifyToken.js]: This feature was already implemented in another file. Check this [file](https://github.com/atlp-rwanda/script-heroes-bn-backend/blob/87a3307d2ce0a9fcba8f9966f9735e0d6488d501/src/middlewares/auth.middleware.js#L5) and see if you can use it.

**Review by dushimeemma — APPROVED** [review]
Good job

**Review by valens10 — COMMENTED** [review]
Your test has failed with this error `The command "npm test" exited with 1.` try to resolve it.

File: src/controllers/completeProfile.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import completeprofileValidations from '../middlewares/validations/completeprofile';
+import db from '../database/models/index';
+
+const completeprofileController = (req, res) => {
+  const { error } = completeprofileValidations(req.body);
+  if (error) {
+    return res.status(400).json({ err: error.details[0].message });
+  }
```
Comment (Habinezajanvier) [inline src/controllers/completeProfile.js]: For concern separation, can you put these lines in other middlewares

File: src/index.js [hunk 5]
```diff
@@ -33,7 +36,9 @@ app.get('/api-docs', swaggerUi.setup(swaggerDocument));
 if (!isProduction) {
   app.use(errorhandler());
 }
-
+app.use('/api', coprofRoute);
+app.use('/api', disprofRoute);
+app.use('/api', updateprofRoute);
```
Comment (Habinezajanvier) [inline src/index.js]: Why can't we use `/routes/index.js` to import all routes and import a single file containing all here in index.js

**Review by elvisiraguha — APPROVED** [review]

**Comment by kipropbrian** [comment]
How are you validating Gender? Any string more than 4 characters work. Can you validate that better.

File: src/controllers/emailAuth.js [hunk 6]
```diff
@@ -36,9 +36,8 @@ class UserController {
         res.status(201).json({ message: res.__('Successfully registered') });
       })
       .catch((error) => {
-        res
-          .status(500)
-          .send({ message: res.__('Something went wrong') });
+        console.log(error.response.body);
```
Comment (kipropbrian) [inline src/controllers/emailAuth.js]: remove console log

Comment (ericus123) [reply src/controllers/emailAuth.js]: done

**Pull request merged**